src/providers/ad/ad_subdomains.c

```diff
@@ -852,9 +886,21 @@ static void ad_get_slave_domain_connect_done(struct tevent_req *subreq)
         return;
     }
 
+    root_dom = ads_get_root_domain(state->be_ctx, state->root_attrs);
+    if (root_dom == NULL) {
+        tevent_req_error(req, ERR_DOMAIN_NOT_FOUND);
+        return;
+    }
+
+    root_sdom = sdap_domain_get(state->opts, root_dom);
+    if (root_sdom == NULL) {
+        tevent_req_error(req, ERR_DOMAIN_NOT_FOUND);
+        return;
+    }
+
     subreq = sdap_search_bases_send(state, state->ev, state->opts,
                                     sdap_id_op_handle(state->sdap_op),
-                                    state->opts->sdom->search_bases,
+                                    root_sdom->search_bases,
```

Comment:
Maybe the search base can already be set as a member of state in ad_get_slave_domain_send() where it is more clear that we want to search in the forest root. If you prefer to call ads_get_root_domain() here I think it would be worth to add a comment that we want to search the slave domains in a DC of the forest root (adding such comment in ad_get_slave_domain_send() woundn't do any harm either :-)

Reply:
No, I think having the state contain as much of the immutable data the request needs is a good idea, I'll fix this.